### PR TITLE
Fix: fill config defaults

### PR DESCRIPTION
**Pull request by nishant95**
This has the fix applied in #302 along with resolution for  #307 

Added test for the bug encountered in #307 as well.

### CODE REVIEW

File: kong/utils_test.go [hunk 1]
```diff
@@ -999,3 +1000,241 @@ func TestFillConsumerGroupPluginDefaults(T *testing.T) {
 		})
 	}
 }
+
+func Test_fillConfigRecord(T *testing.T) {
```
Comment (pmalek) [inline kong/utils_test.go:1004]:
```suggestion
func Test_fillConfigRecord(t *testing.T) {
```

to prevent confusion between and accidentally failing the whole test where only a subtest should fail.

File: kong/utils_test.go [hunk 2]
```diff
@@ -999,3 +1000,241 @@ func TestFillConsumerGroupPluginDefaults(T *testing.T) {
...
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		T.Run(tc.name, func(t *testing.T) {
+			configSchema, err := getConfigSchema(tc.schema)
+			assert.NoError(err)
+			config := fillConfigRecord(configSchema, tc.config)
+			assert.NotNil(config)
```
Comment (pmalek) [inline kong/utils_test.go:1073]:
```suggestion
		t.Run(tc.name, func(t *testing.T) {
			configSchema, err := getConfigSchema(tc.schema)
			require.NoError(t, err)
			config := fillConfigRecord(configSchema, tc.config)
			require.NotNil(t, config)
```

File: kong/utils_test.go [hunk 3]
```diff
@@ -999,3 +1000,241 @@ func TestFillConsumerGroupPluginDefaults(T *testing.T) {
...
+			if diff := cmp.Diff(config, tc.expected); diff != "" {
+				t.Errorf(diff)
+			}
+		})
+	}
+}
+
+func Test_FillPluginsDefaults(T *testing.T) {
```
Comment (pmalek) [inline kong/utils_test.go:1081]:
```suggestion
func Test_FillPluginsDefaults(t *testing.T) {
```

File: kong/utils_test.go [hunk 4]
```diff
@@ -999,3 +1000,241 @@ func TestFillConsumerGroupPluginDefaults(T *testing.T) {
...
+	RunWhenKong(T, ">=2.6.0 <2.8.1")
+	assert := assert.New(T)
+
+	client, err := NewTestClient(nil, nil)
+	assert.NoError(err)
+	assert.NotNil(client)
```
Comment (pmalek) [inline kong/utils_test.go:1087]:
```suggestion
	client, err := NewTestClient(nil, nil)
	require.NoError(t, err)
	require.NotNil(t, client)
```

File: kong/utils_test.go [hunk 5]
```diff
@@ -999,3 +1000,241 @@ func TestFillConsumerGroupPluginDefaults(T *testing.T) {
...
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		T.Run(tc.name, func(t *testing.T) {
+			plugin := tc.plugin
+			fullSchema, err := client.Schemas.Get(defaultCtx, "plugins/statsd")
+			assert.NoError(err)
+			assert.NotNil(fullSchema)
```
Comment (pmalek) [inline kong/utils_test.go:1129]:
```suggestion
		t.Run(tc.name, func(t *testing.T) {
			plugin := tc.plugin
			fullSchema, err := client.Schemas.Get(defaultCtx, "plugins/statsd")
			require.NoError(t, err)
			require.NotNil(t, fullSchema)
```

File: kong/utils_test.go [hunk 6]
```diff
@@ -999,3 +1000,241 @@ func TestFillConsumerGroupPluginDefaults(T *testing.T) {
...
+		})
+	}
+}
+
+func Test_FillPluginsDefaults_RequestTransformer(T *testing.T) {
+	assert := assert.New(T)
+
+	client, err := NewTestClient(nil, nil)
+	assert.NoError(err)
+	assert.NotNil(client)
```
Comment (pmalek) [inline kong/utils_test.go:1148]:
```suggestion
func Test_FillPluginsDefaults_RequestTransformer(t *testing.T) {
	client, err := NewTestClient(nil, nil)
	require.NoError(t, err)
	require.NotNil(t, client)
```

File: kong/utils_test.go [hunk 7]
```diff
@@ -999,3 +1000,241 @@ func TestFillConsumerGroupPluginDefaults(T *testing.T) {
...
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		T.Run(tc.name, func(t *testing.T) {
+			plugin := tc.plugin
+			fullSchema, err := client.Schemas.Get(defaultCtx, "plugins/request-transformer")
+			assert.NoError(err)
+			assert.NotNil(fullSchema)
```
Comment (pmalek) [inline kong/utils_test.go:1230]:
```suggestion
		t.Run(tc.name, func(t *testing.T) {
			plugin := tc.plugin
			fullSchema, err := client.Schemas.Get(defaultCtx, "plugins/request-transformer")
			require.NoError(t, err)
			require.NotNil(t, fullSchema)
```

File: kong/utils_test.go [hunk 8]
```diff
@@ -999,3 +1000,241 @@ func TestFillConsumerGroupPluginDefaults(T *testing.T) {
...
+			assert.NotNil(config)
+			if diff := cmp.Diff(config, tc.expected); diff != "" {
+				t.Errorf(diff)
+			}
+		})
+	}
+}
+
+func Test_FillPluginsDefaults(T *testing.T) {
+	RunWhenKong(T, ">=2.6.0 <2.8.1")
```
[the remark this answers is not in the document]
Comment (nishant95) [reply kong/utils_test.go:1082]: @pmalek 
We wanted to add tests for a default plugin which was impacted by this bug (discussed [here](https://github.com/Kong/go-kong/pull/302#pullrequestreview-1363700741)).

But the schema for this plugin has changed a lot in the newer versions (and we are fetching the schema from Kong in this test), hence I added this version check so that it doesn't fail unnecessarily.
Probably we can handle it differently so that the test is valid for all versions of kong.

[the remark this answers is not in the document]
Comment (pmalek) [reply kong/utils_test.go]: Got it. Perhaps we could add another test that would work with more recent versions of Kong? We can have those either as separate tests named accordingly or as subtests guarded by the `RunWhenKong()` with appropriate versions range. WDYT?

If we leave it like that what would ensure us that users working with plugins in Gateway 3.2 are getting the correct behavior?

[the remark this answers is not in the document]
Comment (nishant95) [reply kong/utils_test.go]: We should definitely add tests for the recent versions. 

One way to do this is to add schema json in the test data itself, instead of fetching it from Kong. But looking at a few other tests it seems we are no longer doing that (Any reason for this?). 

Also, this particular test scenario might get removed if the StatsD plugin changes its schema to remove the array of records from its config in newer versions.

[the remark this answers is not in the document]
Comment (pmalek) [reply kong/utils_test.go]: You're right. In that case: would be a reasonable ask to suggest that we don't rely on the client itself and just test `FillPluginsDefaults` by specifying the schema in the test? This way we're not relying on the Gateway version and we're only concerned about the function that we're testing, not the environment.

[the remark this answers is not in the document]
Comment (nishant95) [reply kong/utils_test.go]: @pmalek I've added schema to the test.

Also, while adding a few more tests I discovered that default values were not being applied for `nil` array of records (working fine for empty array). Added handling for this as well.

File: kong/utils_test.go [hunk 9]
```diff
@@ -999,3 +1203,310 @@ func TestFillConsumerGroupPluginDefaults(T *testing.T) {
 		})
 	}
 }
+
+func Test_fillConfigRecord(t *testing.T) {
+	client, err := NewTestClient(nil, nil)
+	require.NoError(t, err)
+	require.NotNil(t, client)
+
```
Comment (pmalek) [inline kong/utils_test.go:1211]: We don't need this anymore :) 

```suggestion

```

File: kong/utils_test.go [hunk 10]
```diff
@@ -12,8 +12,212 @@ import (
 	"github.com/google/go-cmp/cmp/cmpopts"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"github.com/tidwall/gjson"
 )
 
+var StatsDSchema = `{
```
Comment (pmalek) [inline kong/utils_test.go:18]: I believe this can be made `const`

```suggestion
const StatsDSchema = `{
```

**Review by pmalek — COMMENTED** [review]
Thanks! 👍 Some minor nits.

I'll let @rainest double check when he comes online but looks good to me.

**Review by rainest — APPROVED** [review]
https://github.com/Kong/kubernetes-ingress-controller/pull/3866 trial run integrated into KIC is passing Knative and conformance now, so extra confirmation in integration that the Request Transformer configuration is no longer getting clobbered.

https://github.com/Kong/deck/pull/888 now succeeds as well, where it was failing with 0.39.1

**Comment by nishant95** [comment]
this fixes #307

**Comment by aboudreault** [comment]
@nishant95 @pmalek @rainest looks like this change introduced another bug where an array is completely ignored. we can reproduce with deck and the following:
```
_format_version: "3.0"

plugins:
- name: kafka-log
  enabled: false
  config:
    topic: konnect.auditlogs.gateway-access-logs
    bootstrap_servers:
    - host: broker-1.endpoint
      port: 9096
    - host: broker-1.endpoint
      port: 9096
    - host: broker-1.endpoint
      port: 9096
    authentication:
      strategy: sasl
      mechanism: SCRAM-SHA-512
      user: sasl_user
      password: xxxxxxxx
    security:
      ssl: false
    cluster_name: konnect-msk
```

**Comment by aboudreault** [comment]
ref: https://github.com/Kong/go-kong/pull/333

**Comment by nishant95** [comment]
@aboudreault #309  only adds handling for subconfig of type "array", how does this cause a subconfig of type "set" to fail?
How was this working before this PR was merged?

Also, the [handle_missing_fields](https://github.com/Kong/kong/blob/241c4d0681c7996e8565e278937838f5ec723273/kong/db/schema/init.lua#L1001) in lua code and [FillPluginsDefaults](https://github.com/Kong/go-kong/blob/3ff328d54ff853d7bf24e5f1ab04adb8fd22a317/kong/utils.go#L510) in go-kong seems out of sync. We recently faced an issue in case of non-required subconfig of type record (Testing the fix, yet to raise a PR). 
Kong kept the default as nil but KIC filled the default with a record with nil fields causing KIC to be stuck in sync retry loop and ultimately getting OOMKilled.

**Comment by nishant95** [comment]
@pmalek @rainest @aboudreault 
Fix for #334: #336
